### PR TITLE
First cut at maintenance actions playbook

### PR DESCRIPTION
**Pull request by oneswig**
Rough idea here is:

```
ansible-playbook -i ansible/inventory ansible/cluster-infra.yml -e @config/maintenance.yml
ansible-playbook -i ansible/inventory_maintenance ansible/maintenance.yml -e @config/maintenance.yml
```
The playbooks have some hard-coded data where it isn't obvious how to be smarter.  Also it could definitely do with some checks for idempotency.

It might also be better still to roll everything into different tasks in one playbook, and have reboot_and_wait as a handler.

### CODE REVIEW

File: ansible/roles/mellanox_nic_fw/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+- name: Copy out the Mellanox firmware zipfile
+  copy:
+    src: "{{ mellanox_nic_fw_bin }}"
```
Comment (markgoddard) [inline ansible/roles/mellanox_nic_fw/tasks/main.yml]: nit: Should this be in defaults?

File: ansible/roles/mellanox_nic_fw/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+Reboot a node and wait for it to return
```
Comment (markgoddard) [inline ansible/roles/mellanox_nic_fw/README.md]: Title needs updating

File: ansible/roles/mellanox_nic_fw/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+------------
+
+Mellanox MFT should be installed in the software image.
+
+Role Variables
+--------------
+
+`reboot_down_timeout`: Timeout for node shutdown.  Default is 120 seconds.
+
+`reboot_up_timeout`: Timeout for node restart.  Default is 600 seconds.
```
Comment (markgoddard) [inline ansible/roles/mellanox_nic_fw/README.md]: Role variables need updating.

File: ansible/roles/reboot_and_wait/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+---
+- name: Intiate reboot
+  command: "/usr/sbin/shutdown -r +1"
+  become: True
+
+- name: Ensure the node goes down
+  local_action:
+    module: wait_for
+    host: "{{ ansible_default_ipv4['address'] }}"
```
Comment (markgoddard) [inline ansible/roles/reboot_and_wait/tasks/main.yml]: I normally use ``ansible_host | default(inventory_hostname)``, which I think reflects what ansible uses.

File: ansible/roles/dell_bios_update/tasks/main.yml [hunk 5]
```diff
@@ -7,9 +7,15 @@
...
 
+# Exit codes:
+# 2: Update applied.
+# 3: No update required.
 - name: Execute the BIOS firmware upgrade
   command: "{{ bios_data.dest }} -q"
   become: True
+  register: result
+  failed_when: result.rc not in [2, 3]
+  changed_when: result.rc == 2
```
Comment (oneswig) [inline ansible/roles/dell_bios_update/tasks/main.yml]: Tidy work Mark!

File: ansible/roles/dell_bios_update/tasks/main.yml [hunk 6]
```diff
@@ -23,9 +29,15 @@
...
 
+# Exit codes:
+# 0: Update applied, reboot required.
+# 3: No update required.
 - name: Execute the iDRAC firmware upgrade
   command: "{{ idrac_data.dest }} -q"
   become: True
+  register: result
+  failed_when: result.rc not in [0, 3]
+  changed_when: result.rc == 0
```
Comment (oneswig) [inline ansible/roles/dell_bios_update/tasks/main.yml]: Again, good stuff.

**Pull request merged**